### DISCUSSION TITLE
Differences in output from executable and perl

### DISCUSSION THREAD

**Discussion by stefankinell** [karttapullautin/karttapullautin#4] [General]
@rphlo - I moved the discussion on variations on contour output here.

When I run the same .laz-file on the pullouta.exe on a PC vs running the script om a Mac I get slightly different output. I have noticed this on both the contours as well as 03-contours. 

Here's a picture where red are out2.dxf from a PC with the executable version 20190203 downloaded from https://www.routegadget.net/karttapullautin/ 

The black lines out2.dxf from the perl script with rust parts running on a Mac.

<img width="1030" alt="image" src="https://user-images.githubusercontent.com/59049892/226868406-f8ffb8ad-3aee-4583-aa2f-5df1fda3bc3a.png">

Both directories with output files from "temp" directory are available at https://drive.google.com/drive/folders/1oz-eHR6Q1FbPrg6dKWGlLdVAQSFGpEsF?usp=share_link

It seems the stand alone PC-version is smoothing out the contours a bit more. I also see some differences in form lines, which I think is based on the contours being slightly differently drawn.

**Comment by rphlo** [comment]
This doesn't seems to be major differences... it might be that the windows version is containing slightly more recent code.
Is it a show stopper for you?

**Reply by stefankinell** [reply]
A little bit. I spent some time on setting the .ini-file up so I get the output I like, including how curves look. So I would rather prefere the output from the packaged executable, i.e. PC above.

Did you get the perl-script from Jarkko? We can ask him if he understands why there is a difference. I mean, it is nicer if you build something really good on the last development he did.

**Reply by stefankinell** [reply]
I saw in the perl-script now that it says

$ver = '20190203';

So the perl script are from the same time as the executable. So that should not be what makes the difference. Strange.

**Comment by rphlo** [comment]
@stefankinell What mac software do you use to visualize the dxf files?

**Reply by stefankinell** [reply]
Qgis

**Comment by rphlo** [comment]
I have couple of ideas where the differences come from. Will test the hypothesis in coming days...

**Comment by rphlo** [comment]
@stefankinell Now the rust script behave the same as the Perl version.
The only remaining differences might be due to a bug in the perl version that I corrected in the Rust code.

**Reply by rphlo** [reply]
Recompile the rust version to see the changes

**Reply by stefankinell** [reply]
I downloaded the new version, ran "cargo build --release" and "cp target/release/rusty-pullauta /usr/local/bin/" again.

But still there is the same difference. It is OK output, but the perl/rust version seems to follow the 0,3 contours much more than the executable. Which results in some turns I would have skipped if I draw it manually. So the executable windows version still gives better result. Strange.

**Reply by rphlo** [reply]
I can confirm that the differences are due to the bug in the perl script, I don't think the bug is intentional.
I don't see it as good idea to reproduce the buggy behaviour.

**Reply by rphlo** [reply]
See https://github.com/rphlo/rusty-pullauta/blob/669d08b5ff30cc897669e1c8005f4470913a4da9/pullauta#L4927

the `$$ele` should be `$ele`

**Reply by stefankinell** [reply]
I still get differences in the output. And I am not sure what I am doing wrong.

How do I know that the rust-code is being used?

I tried to execute "rusty-pullauta" instead of "pullauta" but that gave the result of 
"This is rusty karttapulatin... There is no warranty. Use it at your own risk!

Preparing input file
Not implemented further"

Perhaps it is not the intention to execute that directly?

I will try looking into the code and see if the smoothing variable is used at all or if something is fishy there.

**Reply by rphlo** [reply]
if you want to use the original perl script without the rust code, open the pullauta file and search and replace `rusty-pullauta` by `perl pullauta` lines 2170, 2184, 2197 and 2199

**Reply by rphlo** [reply]
Can you share your ini file?

**Reply by stefankinell** [reply]
Sure, but I have been playing around with the standard ini-file today just to make sure it was not my file playing a trick. It is still different.

I have ran simultaneous tests on Mac, with rust / with only perl (does not make a difference) with different smoothing - hardly makes any difference either but little noticeable, and same on PC with the stand alone version. Also on PC with different smoothing.

The biggest difference is between stand alone on one hand (which has in my opinion nicer curves) and perl/rust on the other hand - which is not so fluent with the curves.

I'll reply to Jarkko on twitter as well but I am running out of ideas of what it is that makes it so different. I think I even tried with the same .xyz-file on both platforms to see that it was not las2txt that behaved different.

Have you tried running the stand alone and seeing if you get the same behavior?

**Reply by rphlo** [reply]
I try to dig out my window machine tomorrow to test it...

**Reply by stefankinell** [reply]
Should we agree on a joint -laz-file to use as test object? To simplify testing and comparing? I can pick a part of the open Swedish lidar if you want. Or if you have a preferred file yourself.

**Comment by rphlo** [comment]
@stefankinell 
I just fixed the issue, please fetch latest version.
In summary I had misunderstood a small part of the perl script.

Now Rust+Perl, Perl, and Windows versions behave the exactly the same and produce same output.

**Reply by stefankinell** [reply]
Bravo! The short tests I have done with the new release looks good. Super!

**Comment by stefankinell** [comment]
I have now tested several files and I can concur that I see the same output from the windows stand alone and this new Rust version. So I think we can close this thread.
